subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/capabilities/CapabilitiesUseCasesIntegrationTest.groovy

```diff
@@ -0,0 +1,415 @@
+                            .because('capability cglib is provided by cglib:cglib and cglib:cglib-nodep')
+                            .with(module('cglib:cglib:3.2.5'))
+                    }
+                }
+            }
+        """
+
+        when:
+        repositoryInteractions {
+            'cglib:cglib:3.2.5' {
+                expectResolve()
+            }
+        }
+        run ':checkDeps'
+
+        then:
+        resolve.expectGraph {
+            root(":", ":test:") {
+                module('cglib:cglib:3.2.5').byReason('capability cglib is provided by cglib:cglib and cglib:cglib-nodep')
+                edge('cglib:cglib-nodep:3.2.5', 'cglib:cglib:3.2.5').byReason('capability cglib is provided by cglib:cglib and cglib:cglib-nodep')
+            }
+        }
+
+    }
+
+    /**
+     * This use case corresponds to the case where a library maintainer publishes a library
+     * either as individual modules, or using a fatjar. It's illegal to have both the fat jar and
+     * individual modules on the classpath, so we need a way to declare that we prefer to use
+     * the fat jar.
+     *
+     * This is from the consumer point of view, fixing the fact the library doesn't declare capabilities.
+     */
+    def "can select groovy-all over individual groovy-whatever"() {
+        given:
+        repository {
+            'org.apache:groovy:1.0'()
+            'org.apache:groovy-json:1.0'()
+            'org.apache:groovy-xml:1.0'()
+            'org.apache:groovy-all:1.0'()
+
+            'org:a:1.0' {
+                dependsOn 'org.apache:groovy:1.0'
+                dependsOn 'org.apache:groovy-json:1.0'
+            }
+            'org:b:1.0' {
+                dependsOn 'org.apache:groovy-all:1.0'
+            }
+        }
+
+        buildFile << """
+            dependencies {
+               conf "org:a:1.0"
+               conf "org:b:1.0"
+            
+               components {
+                  withModule('org.apache:groovy') {
+                     allVariants {
+                        withCapabilities {
+                            addCapability('org.apache', 'groovy', '1.0')
+                        }
+                     }
+                  }
+                  withModule('org.apache:groovy-json') {
+                     allVariants {
+                        withCapabilities {
+                            addCapability('org.apache', 'groovy-json', '1.0')
+                        }
+                     }
+                  }
+                  withModule('org.apache:groovy-all') {
+                     allVariants {
+                        withCapabilities {
+                            addCapability('org.apache', 'groovy', '1.0')
+                            addCapability('org.apache', 'groovy-json', '1.0')
+                        }
+                     }
+                  }
+               }               
+
+               // solution
```

Comment:
Again, I think first resolving with failure would make this a better demonstration.

Reply:
+1